Log results of each plugin in scheduler framework

wgliang opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind cleanup


**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #83401

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Log results of each plugin in scheduler framework
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

wgliang committed:
Log results of each plugin in scheduler framework

wgliang committed:
fix RunPreFilter panic error when pod is nil

alculquicondor commented:
/assign @liu-cong

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -210,6 +214,7 @@ func (f *framework) RunPreFilterPlugins(
 			klog.Error(msg)
 			return NewStatus(Error, msg)
 		}
+		klog.V(10).Infof("running %q prefilter plugin for pod %q successfully", pl.Name(), pod.Name)
```
it's easier to just log the status in my opinion. This makes sure the status is always logged and you don't need to worry about whether the status is success or not. Makes the control flow easier.

wgliang replied:
Added status message.

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -475,6 +523,8 @@ func (f *framework) RunPermitPlugins(
 				return NewStatus(Error, msg)
 			}
+		} else {
+			klog.V(10).Infof("running %q permit plugin for pod %q successfully", pl.Name(), pod.Name)
```
easier to just log the status and avoid the `else` block

wgliang replied:
SG

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -502,7 +552,10 @@ func (f *framework) RunPermitPlugins(
+			klog.V(10).Infof("running permit plugins for pod %q successfully", pod.Name)
 		}
+	} else {
+		klog.V(10).Infof("running permit plugins for pod %q successfully", pod.Name)
```
we shouldn't need this.

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -502,7 +552,10 @@ func (f *framework) RunPermitPlugins(
 				klog.Error(msg)
 				return NewStatus(Error, msg)
 			}
+			klog.V(10).Infof("running permit plugins for pod %q successfully", pod.Name)
```
the message should say something like "wait complete for permit plugins"

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework_test.go:
```diff
@@ -494,16 +494,16 @@ func TestPreFilterPlugins(t *testing.T) {
 		f.RunPreFilterExtensionRemovePod(nil, nil, nil, nil)
 
-		if preFilter1.PreFilterCalled != 1 {
+		if preFilter1.PreFilterCalled != 0 {
```
why making this change?

wgliang replied:
This is related with https://github.com/kubernetes/kubernetes/pull/83538#discussion_r332316045

wgliang commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -198,6 +198,10 @@ func (f *framework) QueueSortFunc() LessFunc {
 	state *CycleState, pod *v1.Pod) *Status {
+	if pod == nil {
+		return nil
+	}
```
Yes, this is because panic does occur, in unit testing:

you can run

`make test WHAT=./pkg/scheduler/framework/v1alpha1/`

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x178eb08]

goroutine 79 [running]:
testing.tRunner.func1(0xc0001e7000)
	/usr/local/Cellar/go/1.13.1/libexec/src/testing/testing.go:874 +0x3a3
panic(0x1859120, 0x203a5f0)
	/usr/local/Cellar/go/1.13.1/libexec/src/runtime/panic.go:679 +0x1b2
k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1.(*framework).RunPreFilterPlugins(0xc000200dc0, 0x0, 0x0, 0x0)
	/Users/wangguoliang/Documents/mos/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1/framework.go:214 +0x138
k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1.TestPreFilterPlugins.func3(0xc0001e7000)
	/Users/wangguoliang/Documents/mos/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1/framework_test.go:493 +0x10b
testing.tRunner(0xc0001e7000, 0xc0001e3080)
	/usr/local/Cellar/go/1.13.1/libexec/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/Cellar/go/1.13.1/libexec/src/testing/testing.go:960 +0x350
FAIL	k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1	0.021s
```

wgliang replied:
FYI:
https://github.com/kubernetes/kubernetes/blob/22ed5c2fe2a4f5da37ecedb58709a4b778b4658e/pkg/scheduler/framework/v1alpha1/framework_test.go#L493

liu-cong replied:
Why does the test call RunPrefilterPlugins with nil pod?

wgliang replied:
FYI:https://github.com/kubernetes/kubernetes/commit/37b9e6d1eadb

ahg-g replied:
No.

liu-cong replied:
@wgliang , can we update the tests and get rid of the nil checks here? Thanks.

wgliang replied:
done

wgliang commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -272,6 +293,7 @@ func (f *framework) RunFilterPlugins(state *CycleState,
 			}
 			return status
 		}
+		klog.V(10).Infof("running %q filter plugin for pod %q successfully", pl.Name(), pod.Name)
```
SGTM, but I think the node name is enough (nodeInfo has too much information). What do you think this? @liu-cong

wgliang replied:
done

wgliang commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -210,6 +214,7 @@ func (f *framework) RunPreFilterPlugins(
 			klog.Error(msg)
 			return NewStatus(Error, msg)
 		}
+		klog.V(10).Infof("running %q prefilter plugin for pod %q successfully", pl.Name(), pod.Name)
```
done

wgliang commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -372,6 +400,14 @@ func (f *framework) RunScorePlugins(state *CycleState, pod *v1.Pod, nodes []*v1.
 		return nil, NewStatus(Error, msg)
 	}
+	if klog.V(10) {
+		for pluginName, nodeScores := range pluginToNodeScores {
```
done

wgliang commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -198,6 +198,10 @@ func (f *framework) QueueSortFunc() LessFunc {
+	if pod == nil {
+		return nil
+	}
+	klog.V(10).Infof("RunPreFilterPlugins: ")
```
done

liu-cong reviewed (commented):
Can you mark resolved comments as "Resolved"? Thanks!

wgliang commented:
/test pull-kubernetes-e2e-gce

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -331,6 +363,8 @@ func (f *framework) RunScorePlugins(state *CycleState, pod *v1.Pod, nodes []*v1.
 		return nil, NewStatus(Error, msg)
 	}
 
+	klog.V(10).Infof("%+v", pluginToNodeScores)
```
nit: please add "pluginToNodeScores after Score:%+v", and similarly "pluginToNodeScores after NormalizeScore:%+v", "pluginToNodeScores after ApplyWeights:%+v"

wgliang replied:
done

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -210,6 +214,7 @@ func (f *framework) RunPreFilterPlugins(
 			klog.Error(msg)
 			return NewStatus(Error, msg)
 		}
+		klog.V(10).Infof("Running %q prefilter plugin for pod %q: %v", pl.Name(), pod.Name, status.Message())
```
nit: Consider moving this before the `if !status.IsSuccess()`, otherwise the status will always be success here. Same comment for other occurrences.

wgliang replied:
In fact, this is to take into account that will print information duplicates with `klog.Error(msg)`.

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -426,20 +471,24 @@ func (f *framework) RunPostBindPlugins(
 // returns the error. In such case, pod will not be scheduled.
 func (f *framework) RunReservePlugins(
 	state *CycleState, pod *v1.Pod, nodeName string) *Status {
+	klog.V(10).Infof("RunBindPlugins for pod %q: ", pod)
```
RunReservePlugins

wgliang replied:
done

liu-cong commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -502,6 +553,7 @@ func (f *framework) RunPermitPlugins(
 				klog.Error(msg)
 				return NewStatus(Error, msg)
 			}
+			klog.V(10).Infof("wait complete for permit plugins for pod %q: %v", pod.Name, statusCode)
```
nit: let's also log the permit plugin name. 
nit: no need to log statusCode (it's Wait).

wgliang replied:
There is no specified or explicit permit plugin here, here is the wait for all the permit plugin to complete.

wgliang replied:
done

liu-cong commented:
/lgtm

Thanks Leon!

/assign ahg-g

wgliang commented:
/ping @ahg-g

wgliang commented:
/priority important-soon

liu-cong commented:
can you rebase

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -331,6 +348,8 @@ func (f *framework) RunScorePlugins(state *CycleState, pod *v1.Pod, nodes []*v1.
 		return nil, NewStatus(Error, msg)
 	}
 
+	klog.V(10).Infof("pluginToNodeScores after Score: %+v", pluginToNodeScores)
```
this is going to be massive for large clusters. Please remove it.

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -220,11 +222,15 @@ func (f *framework) RunPreFilterPlugins(
 // status other than Success.
 func (f *framework) RunPreFilterExtensionAddPod(state *CycleState, podToSchedule *v1.Pod,
 	podToAdd *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status {
+	klog.V(10).Infof("RunPreFilterPlugins for pod %q: ", podToSchedule)
```
ditto, not useful.

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -240,11 +246,15 @@ func (f *framework) RunPreFilterExtensionAddPod(state *CycleState, podToSchedule
 // status other than Success.
 func (f *framework) RunPreFilterExtensionRemovePod(state *CycleState, podToSchedule *v1.Pod,
 	podToRemove *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status {
+	klog.V(10).Infof("RunPreFilterPlugins for pod %q: ", podToSchedule)
```
ditto, not useful.

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -286,8 +300,10 @@ func (f *framework) RunPostFilterPlugins(
 	nodes []*v1.Node,
 	filteredNodesStatuses NodeToStatusMap,
 ) *Status {
+	klog.V(10).Infof("RunPostFilterPlugins: ")
```
ditto, not useful.

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -373,6 +394,8 @@ func (f *framework) RunScorePlugins(state *CycleState, pod *v1.Pod, nodes []*v1.
 		return nil, NewStatus(Error, msg)
 	}
 
+	klog.V(10).Infof("pluginToNodeScores after ApplyWeights: %+v", pluginToNodeScores)
```
This is going to be massive for large clusters. Please remove it.

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -426,20 +456,24 @@ func (f *framework) RunPostBindPlugins(
 // RunUnreservePlugins runs the set of configured unreserve plugins.
 func (f *framework) RunUnreservePlugins(
 	state *CycleState, pod *v1.Pod, nodeName string) {
+	klog.V(10).Infof("RunUnreservePlugins for pod %q: ", pod)
```
ditto, not useful.

ahg-g reviewed (commented):
I put some 19 comments on this PR, I am very skeptical that this is going to be useful, logging for each pod for each node on each extension point is massive, even if it was enabled temporarily. I don' think this should be merged.

wgliang commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -397,10 +423,13 @@ func (f *framework) RunBindPlugins(state *CycleState, pod *v1.Pod, nodeName stri
+	klog.V(10).Infof("RunBindPlugins for pod %q: ", pod)
 	var status *Status
-	for _, bp := range f.bindPlugins {
+	for pluginName, bp := range f.bindPlugins {
```
done

wgliang commented:
> > I put some 19 comments on this PR, I am very skeptical that this is going to be useful, logging for each pod for each node on each extension point is massive, even if it was enabled temporarily. I don' think this should be merged.
> 
> if we want more insights from logs, we can do that outside the framework: log a line after running all filters on all nodes for example.

@ahg-g Thanks for your review. I've updated and rebased it.

wgliang commented on pkg/scheduler/framework/v1alpha1/framework_test.go:
```diff
@@ -586,7 +594,7 @@ func TestRecordingMetrics(t *testing.T) {
 		{
 			name:               "Filter - Success",
-			action:             func(f Framework) { f.RunFilterPlugins(context.Background(), nil, pod, nil) },
+			action:             func(f Framework) { f.RunFilterPlugins(context.Background(), nil, pod, nodeInfo) },
```
This is to solve the panic issue.

wgliang commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -334,10 +339,11 @@ func (f *framework) RunFilterPlugins(
-				errMsg := fmt.Sprintf("error while running %q filter plugin for pod %q: %v",
-					pl.Name(), pod.Name, status.Message())
+				errMsg := fmt.Sprintf("error while running %q filter plugin for pod %q on node %v: %v",
+					pl.Name(), pod.Name, nodeInfo.Node().Name, status.Message())
```
Print `nodeInfo.Node().Name`
